File: C++/数据结构/LazySeg.cpp
struct Tag {
    ll add = 0;  // 懒标记初始值

    Tag(ll add = 0) : add(add) {}

    bool empty() const { return add == 0; }

    void apply(const Tag& t) {
        add += t.add;
    }  // 合并懒标记：先已有操作，再做t
};

struct Info {
    ll sum = 0;

    Info(ll sum = 0) : sum(sum) {}

    void apply(const Tag& t, int l, int r) {
        sum += t.add * (r - l + 1);
    }  // 把懒标记作用到当前节点
};

Info operator+(const Info& a, const Info& b) {
    return {a.sum + b.sum};
}  // 合并两个Info

bool operator<(const Info& a, const Info& b) {
    return a.sum < b.sum;
}  // 线段树二分用，不需要时可删

template <typename Info, typename Tag>
class LazySegmentTree {
    int n;
    vector<Info> info;
    vector<Tag> tag;

    void apply(int node, int l, int r, const Tag& v) {
        info[node].apply(v, l, r);
        tag[node].apply(v);
    }

    void pushdown(int node, int l, int r) {
        if (tag[node].empty()) return;
        int m = (l + r) >> 1;
        apply(node << 1, l, m, tag[node]);
        apply(node << 1 | 1, m + 1, r, tag[node]);
        tag[node] = Tag();
    }  // 把当前节点的懒标记下传

    void maintain(int node) {
        info[node] = info[node << 1] + info[node << 1 | 1];
    }

    void build(const vector<Info>& a, int node, int l, int r) {
        if (l == r) {
            info[node] = a[l];
            return;
        }
        int m = (l + r) >> 1;
        build(a, node << 1, l, m);
        build(a, node << 1 | 1, m + 1, r);
        maintain(node);
    }  // 建树，复杂度O(n)

    void update(int node, int l, int r, int ql, int qr, const Tag& v) {
        if (ql <= l && r <= qr) {
            apply(node, l, r, v);
            return;
        }
        pushdown(node, l, r);
        int m = (l + r) >> 1;
        if (ql <= m) update(node << 1, l, m, ql, qr, v);
        if (qr > m) update(node << 1 | 1, m + 1, r, ql, qr, v);
        maintain(node);
    }  // 区间更新[ql,qr]

    void assign(int node, int l, int r, int p, const Info& v) {
        if (l == r) {
            info[node] = v;
            tag[node] = Tag();
            return;
        }
        pushdown(node, l, r);
        int m = (l + r) >> 1;
        if (p <= m)
            assign(node << 1, l, m, p, v);
        else
            assign(node << 1 | 1, m + 1, r, p, v);
        maintain(node);
    }  // 单点赋值

    Info query(int node, int l, int r, int ql, int qr) {
        if (ql <= l && r <= qr) return info[node];
        pushdown(node, l, r);
        int m = (l + r) >> 1;
        if (qr <= m) return query(node << 1, l, m, ql, qr);
        if (ql > m) return query(node << 1 | 1, m + 1, r, ql, qr);
        return query(node << 1, l, m, ql, qr) + query(node << 1 | 1, m + 1, r, ql, qr);
    }  // 区间查找

    template <typename F>
    int find_first(int node, int l, int r, int ql, int qr, F&& check) {
        if (r < ql || l > qr || !check(info[node])) return -1;
        if (l == r) return l;
        pushdown(node, l, r);
        int m = (l + r) >> 1;
        int res = find_first(node << 1, l, m, ql, qr, check);
        if (res != -1) return res;
        return find_first(node << 1 | 1, m + 1, r, ql, qr, check);
    }  // 若遇到固定左端点的情况，需要使用全局变量（或者传入引用）记录前缀分段最大值，加一个被待求区间完全覆盖的剪枝

    template <typename F>
    int find_last(int node, int l, int r, int ql, int qr, F&& check) {
        if (r < ql || l > qr || !check(info[node])) return -1;
        if (l == r) return l;
        pushdown(node, l, r);
        int m = (l + r) >> 1;
        int res = find_last(node << 1 | 1, m + 1, r, ql, qr, check);
        if (res != -1) return res;
        return find_last(node << 1, l, m, ql, qr, check);
    }

public:
    LazySegmentTree(int n, Info init_val = Info()) : LazySegmentTree(vector<Info>(n, init_val)) {}
    // 维护下标为[0,n-1],初始值为init_val的区间，或者数组a
    LazySegmentTree(const vector<Info>& a) : n(sz(a)), info(2 << bit_width((unsigned)sz(a) - 1)), tag(2 << bit_width((unsigned)sz(a) - 1)) {
        build(a, 1, 0, n - 1);
    }
    // 更新[ql,qr]为f
    void update(int ql, int qr, const Tag& v) { update(1, 0, n - 1, ql, qr, v); }
    // 单点赋值a[p]=v
    void assign(int p, const Info& v) { assign(1, 0, n - 1, p, v); }
    // 区间查询[ql,qr]
    Info query(int ql, int qr) { return query(1, 0, n - 1, ql, qr); }

    template <typename F>
    int find_first(int ql, int qr, F&& check) {
        return find_first(1, 0, n - 1, ql, qr, check);
    }  // 查询[ql,qr]中第一个满足条件的下标

    template <typename F>
    int find_last(int ql, int qr, F&& check) {
        return find_last(1, 0, n - 1, ql, qr, check);
    }  // 查询[ql,qr]中最后一个满足条件的下标

    int find_first(int ql, int qr, const Info& val) {
        return find_first(ql, qr, [&](const Info& x) { return !(x < val); });
    }

    int find_last(int ql, int qr, const Info& val) {
        return find_last(ql, qr, [&](const Info& x) { return !(x < val); });
    }
};
// 注：懒标记线段树无论做什么都需要pushdown
// 此时其它与线段树二分同

// 双标记，注意顺序。使用同一个LazySegmentTree<ModInfo, ModTag>
struct ModTag {
    static constexpr int MOD = 1e9 + 7;
    ll add = 0;
    ll mul = 1;

    ModTag(ll add = 0, ll mul = 1) : add(add), mul(mul) {}

    bool empty() const { return add == 0 && mul == 1; }

    void apply(const ModTag& t) {
        add = (add * t.mul + t.add) % MOD;
        mul = mul * t.mul % MOD;
    }  // 先已有操作，再做t
};

struct ModInfo {
    static constexpr int MOD = 1e9 + 7;
    ll sum = 0;

    ModInfo(ll sum = 0) : sum(sum) {}

    void apply(const ModTag& t, int l, int r) {
        sum = (sum * t.mul + t.add * ((r - l + 1) % MOD)) % MOD;
    }
};

ModInfo operator+(const ModInfo& a, const ModInfo& b) {
    return {(a.sum + b.sum) % ModInfo::MOD};
}

// 区间加：tree.update(l, r, {x, 1})
// 区间乘：tree.update(l, r, {0, x})

// LC850:矩形面积并（扫描线）
const ll MOD = 1e9 + 7;
class SegmentTree {
private:
    struct Node {
        int l, r;
        int min_cover_len = 0;  // 区间内被覆盖的最小次数
        int min_cover = 0;      // 区间内为最小次数的区间长度
        int todo = 0;           // 懒标记
    };

    vector<Node> seg;

    void maintain(int o) {
        Node& lo = seg[o << 1];
        Node& ro = seg[(o << 1) | 1];
        int mn = min(lo.min_cover, ro.min_cover);
        seg[o].min_cover = mn;
        seg[o].min_cover_len = (lo.min_cover == mn ? lo.min_cover_len : 0) + (ro.min_cover == mn ? ro.min_cover_len : 0);
    }  // 根据左右儿子的信息，更新当前节点的信息

    void do_(int o, int v) {
        seg[o].min_cover += v;
        seg[o].todo += v;
    }  // 仅更新节点信息，不下传懒标记

    void pushdown(int o) {
        int& v = seg[o].todo;
        if (v) {
            do_(o << 1, v);
            do_(o << 1 | 1, v);
            v = 0;
        }
    }  // 下传懒标记

    void build(vi& xs, int o, int l, int r) {
        seg[o].l = l;
        seg[o].r = r;
        if (l == r) {
            seg[o].min_cover_len = xs[l + 1] - xs[l];
            return;
        }
        int m = (l + r) >> 1;
        build(xs, o << 1, l, m);
        build(xs, o << 1 | 1, m + 1, r);
        maintain(o);
        return;
    }

    void update(int o, int l, int r, int v) {
        if (l <= seg[o].l && seg[o].r <= r) {
            do_(o, v);
            return;
        }
        pushdown(o);
        int m = (seg[o].l + seg[o].r) >> 1;
        if (l <= m) update(o << 1, l, r, v);
        if (m < r) update(o << 1 | 1, l, r, v);
        maintain(o);
    }

public:
    SegmentTree(vi& xs) {
        unsigned n = sz(xs) - 1;  // 有这么多个差值
        seg.resize(2 << bit_width(n - 1));
        build(xs, 1, 0, n - 1);  // 根节点是1
    }

    void update(int l, int r, int v) { update(1, l, r, v); }

    int get_uncovered_length() { return seg[1].min_cover ? 0 : seg[1].min_cover_len; }
};// 这个模板用的是左闭右开
class Solution {
public:
    int rectangleArea(vector<vector<int>>& rectangles) {
        int n = sz(rectangles);
        vi xs;
        struct Event {
            int y, lx, rx, d;
        };
        vector<Event> tem;
        for (auto& p : rectangles) {
            int lx = p[0], rx = p[2], ly = p[1], ry = p[3];
            xs.push_back(lx);
            xs.push_back(rx);
            tem.emplace_back(ly, lx, rx, 1);
            tem.emplace_back(ry, lx, rx, -1);
        }
        ranges::sort(xs);
        xs.erase(unique(all(xs)), xs.end());
        SegmentTree tree(xs);
        sort(all(tem), [&](const Event& x, const Event& y) { return x.y < y.y; });
        int m = sz(tem);
        ll ans = 0;
        rep(i, 0, m - 2) {
            auto& [y, lx, rx, d] = tem[i];
            int l = ranges::lower_bound(xs, lx) - xs.begin();
            int r = ranges::lower_bound(xs, rx) - xs.begin() - 1;  // 注意点和区间的对应关系
            tree.update(l, r, d);
            int sum = xs.back() - xs[0] - tree.get_uncovered_length();
            ans += 1LL * sum * (tem[i + 1].y - y);
            ans %= MOD;
        }
        return ans % MOD;
    }
};

// 动态开点懒标记线段树
struct Tag {
    ll add = 0;  // 懒标记初始值

    Tag(ll add = 0) : add(add) {}

    bool empty() const { return add == 0; }

    void apply(const Tag& t) {
        add += t.add;
    }  // 合并懒标记：先已有操作，再做t
};

struct Info {
    ll sum = 0;

    Info(ll sum = 0) : sum(sum) {}

    void apply(const Tag& t, int l, int r) {
        sum += t.add * (r - l + 1);
    }  // 把懒标记作用到当前节点
};

Info operator+(const Info& a, const Info& b) {
    return {a.sum + b.sum};
}  // 合并两个Info

bool operator<(const Info& a, const Info& b) {
    return a.sum < b.sum;
}  // 线段树二分用，不需要时可删

template <typename Info, typename Tag>
class DynamicLazySegmentTree {
    struct Node {
        int left = -1, right = -1;
        Info info = Info();
        Tag tag = Tag();
    };

    int n;
    vector<Node> tree;

    int new_node() {
        tree.emplace_back();
        return tree.size() - 1;
    }

    Info get_info(int node) const { return node == -1 ? Info() : tree[node].info; }

    int left_child(int node) {
        if (tree[node].left == -1) {
            int child = new_node();
            tree[node].left = child;
        }
        return tree[node].left;
    }

    int right_child(int node) {
        if (tree[node].right == -1) {
            int child = new_node();
            tree[node].right = child;
        }
        return tree[node].right;
    }

    void apply(int node, int l, int r, const Tag& v) {
        tree[node].info.apply(v, l, r);
        tree[node].tag.apply(v);
    }

    void pushdown(int node, int l, int r) {
        if (tree[node].tag.empty() || l == r) return;
        Tag t = tree[node].tag;
        int m = (l + r) >> 1;
        int left = left_child(node);
        int right = right_child(node);
        apply(left, l, m, t);
        apply(right, m + 1, r, t);
        tree[node].tag = Tag();
    }  // 把当前节点的懒标记下传

    void maintain(int node) {
        tree[node].info = get_info(tree[node].left) + get_info(tree[node].right);
    }

    void update(int node, int l, int r, int ql, int qr, const Tag& v) {
        if (ql <= l && r <= qr) {
            apply(node, l, r, v);
            return;
        }
        pushdown(node, l, r);
        int m = (l + r) >> 1;
        if (ql <= m) update(left_child(node), l, m, ql, qr, v);
        if (qr > m) update(right_child(node), m + 1, r, ql, qr, v);
        maintain(node);
    }  // 区间更新[ql,qr]

    void assign(int node, int l, int r, int p, const Info& v) {
        if (l == r) {
            tree[node].info = v;
            tree[node].tag = Tag();
            return;
        }
        pushdown(node, l, r);
        int m = (l + r) >> 1;
        if (p <= m)
            assign(left_child(node), l, m, p, v);
        else
            assign(right_child(node), m + 1, r, p, v);
        maintain(node);
    }  // 单点赋值

    Info query(int node, int l, int r, int ql, int qr) {
        if (node == -1) return Info();
        if (ql <= l && r <= qr) return tree[node].info;
        pushdown(node, l, r);
        int m = (l + r) >> 1;
        if (qr <= m) return query(tree[node].left, l, m, ql, qr);
        if (ql > m) return query(tree[node].right, m + 1, r, ql, qr);
        return query(tree[node].left, l, m, ql, qr) + query(tree[node].right, m + 1, r, ql, qr);
    }  // 区间查询

    template <typename F>
    int find_first(int node, int l, int r, int ql, int qr, F&& check) {
        if (r < ql || l > qr || !check(get_info(node))) return -1;
        if (l == r) return l;
        if (node != -1) pushdown(node, l, r);
        int m = (l + r) >> 1;
        int left = node == -1 ? -1 : tree[node].left;
        int right = node == -1 ? -1 : tree[node].right;
        int res = find_first(left, l, m, ql, qr, check);
        if (res != -1) return res;
        return find_first(right, m + 1, r, ql, qr, check);
    }

    template <typename F>
    int find_last(int node, int l, int r, int ql, int qr, F&& check) {
        if (r < ql || l > qr || !check(get_info(node))) return -1;
        if (l == r) return l;
        if (node != -1) pushdown(node, l, r);
        int m = (l + r) >> 1;
        int left = node == -1 ? -1 : tree[node].left;
        int right = node == -1 ? -1 : tree[node].right;
        int res = find_last(right, m + 1, r, ql, qr, check);
        if (res != -1) return res;
        return find_last(left, l, m, ql, qr, check);
    }

public:
    DynamicLazySegmentTree(int n) : n(n), tree(1) {}

    // 更新[ql,qr]
    void update(int ql, int qr, const Tag& v) { update(0, 0, n - 1, ql, qr, v); }

    // 单点赋值a[p]=v
    void assign(int p, const Info& v) { assign(0, 0, n - 1, p, v); }

    // 区间查询[ql,qr]
    Info query(int ql, int qr) { return query(0, 0, n - 1, ql, qr); }

    template <typename F>
    int find_first(int ql, int qr, const F& check) {
        return find_first(0, 0, n - 1, ql, qr, check);
    }  // 查询[ql,qr]中第一个满足条件的下标

    template <typename F>
    int find_last(int ql, int qr, const F& check) {
        return find_last(0, 0, n - 1, ql, qr, check);
    }  // 查询[ql,qr]中最后一个满足条件的下标

    int find_first(int ql, int qr, const Info& val) {
        return find_first(ql, qr, [&](const Info& x) { return !(x < val); });
    }

    int find_last(int ql, int qr, const Info& val) {
        return find_last(ql, qr, [&](const Info& x) { return !(x < val); });
    }
};

/*
用法：

DynamicLazySegmentTree<Info, Tag> tree(1'000'000'001); // 维护[0, 1e9]
tree.update(l, r, Tag(add));                           // 区间修改
tree.assign(i, Info(val));                             // 单点赋值
ll res = tree.query(l, r).sum;                         // 区间查询
int first = tree.find_first(l, r, Info(val));          // 第一个满足条件的位置
int last = tree.find_last(l, r, Info(val));            // 最后一个满足条件的位置

每次操作时间复杂度O(log n)，空间复杂度O(操作次数 * log n)。
*/

File: Python/数据结构/LazySeg.py
from copy import copy


class Tag:
    __slots__ = ("add",)

    def __init__(self, add: int = 0) -> None:
        self.add = add

    def empty(self) -> bool:
        return self.add == 0

    def apply(self, other: "Tag") -> None:
        self.add += other.add  # 先做已有操作，再做 other


class Info:
    __slots__ = ("mx", "sum")

    def __init__(self, sum: int = 0, mx=None) -> None:
        self.sum = sum
        self.mx = sum if mx is None else mx

    def apply(self, tag: Tag, l: int, r: int) -> None:
        self.sum += tag.add * (r - l + 1)
        self.mx += tag.add

    def __add__(self, other: "Info") -> "Info":
        return Info(self.sum + other.sum, max(self.mx, other.mx))

    def __lt__(self, other: "Info") -> bool:
        return self.mx < other.mx  # 线段树二分用，不需要时可删


class LazySegmentTree:
    def __init__(self, a, init_val=None) -> None:
        if isinstance(a, int):
            init_val = Info() if init_val is None else init_val
            a = [copy(init_val) for _ in range(a)]
        self.n = len(a)
        self.info = [Info() for _ in range(4 * self.n)]
        self.tag = [Tag() for _ in range(4 * self.n)]
        self._build(a, 1, 0, self.n - 1)

    def _apply(self, node: int, l: int, r: int, val: Tag) -> None:
        self.info[node].apply(val, l, r)
        self.tag[node].apply(val)

    def _pushdown(self, node: int, l: int, r: int) -> None:
        if self.tag[node].empty():
            return
        m = (l + r) >> 1
        self._apply(node << 1, l, m, self.tag[node])
        self._apply(node << 1 | 1, m + 1, r, self.tag[node])
        self.tag[node] = Tag()

    def _maintain(self, node: int) -> None:
        self.info[node] = self.info[node << 1] + self.info[node << 1 | 1]

    def _build(self, a: list[Info], node: int, l: int, r: int) -> None:
        if l == r:
            self.info[node] = copy(a[l])
            return
        m = (l + r) >> 1
        self._build(a, node << 1, l, m)
        self._build(a, node << 1 | 1, m + 1, r)
        self._maintain(node)

    def _update(self, node: int, l: int, r: int, ql: int, qr: int, val: Tag) -> None:
        if ql <= l and r <= qr:
            self._apply(node, l, r, val)
            return
        self._pushdown(node, l, r)
        m = (l + r) >> 1
        if ql <= m:
            self._update(node << 1, l, m, ql, qr, val)
        if qr > m:
            self._update(node << 1 | 1, m + 1, r, ql, qr, val)
        self._maintain(node)

    def _assign(self, node: int, l: int, r: int, i: int, val: Info) -> None:
        if l == r:
            self.info[node] = copy(val)
            self.tag[node] = Tag()
            return
        self._pushdown(node, l, r)
        m = (l + r) >> 1
        if i <= m:
            self._assign(node << 1, l, m, i, val)
        else:
            self._assign(node << 1 | 1, m + 1, r, i, val)
        self._maintain(node)

    def _query(self, node: int, l: int, r: int, ql: int, qr: int) -> Info:
        if ql <= l and r <= qr:
            return copy(self.info[node])
        self._pushdown(node, l, r)
        m = (l + r) >> 1
        if qr <= m:
            return self._query(node << 1, l, m, ql, qr)
        if ql > m:
            return self._query(node << 1 | 1, m + 1, r, ql, qr)
        return self._query(node << 1, l, m, ql, qr) + self._query(
            node << 1 | 1, m + 1, r, ql, qr
        )

    def _find_first(
        self, node: int, l: int, r: int, ql: int, qr: int, val: Info
    ) -> int:
        if r < ql or qr < l or self.info[node] < val:
            return -1
        if l == r:
            return l
        self._pushdown(node, l, r)
        m = (l + r) >> 1
        res = self._find_first(node << 1, l, m, ql, qr, val)
        if res != -1:
            return res
        return self._find_first(node << 1 | 1, m + 1, r, ql, qr, val)

    def _find_last(self, node: int, l: int, r: int, ql: int, qr: int, val: Info) -> int:
        if r < ql or qr < l or self.info[node] < val:
            return -1
        if l == r:
            return l
        self._pushdown(node, l, r)
        m = (l + r) >> 1
        res = self._find_last(node << 1 | 1, m + 1, r, ql, qr, val)
        if res != -1:
            return res
        return self._find_last(node << 1, l, m, ql, qr, val)

    def update(self, ql: int, qr: int, val: Tag) -> None:
        self._update(1, 0, self.n - 1, ql, qr, val)  # 修改闭区间 [ql, qr]

    def assign(self, i: int, val: Info) -> None:
        self._assign(1, 0, self.n - 1, i, val)

    def query(self, ql: int, qr: int) -> Info:
        return self._query(1, 0, self.n - 1, ql, qr)

    def find_first(self, ql: int, qr: int, val: Info) -> int:
        return self._find_first(1, 0, self.n - 1, ql, qr, val)

    def find_last(self, ql: int, qr: int, val: Info) -> int:
        return self._find_last(1, 0, self.n - 1, ql, qr, val)


# 用法（n > 0；下标均为 0-based；区间均为闭区间）
# tree = LazySegmentTree([Info(x) for x in a])
# tree = LazySegmentTree(n, Info(0))
# tree.update(l, r, Tag(x))                 # 区间加 x
# tree.assign(i, Info(x))                   # 单点赋值
# res = tree.query(l, r)                    # res.sum / res.mx
# p = tree.find_first(l, r, Info(x))        # 第一个 a[p] >= x，不存在返回 -1
# p = tree.find_last(l, r, Info(x))         # 最后一个 a[p] >= x，不存在返回 -1
# 换题时修改 Info.apply、Info.__add__、Tag.empty 和 Tag.apply
